include/swift/Basic/Located.h

```diff
@@ -0,0 +1,57 @@
+//===--- Located.h - Source Location and Associated Value ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Provides a currency data type Located<T> that should be used instead
+// of std::pair<T, SourceLoc>.
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef SWIFT_BASIC_LOCATED_H
+#define SWIFT_BASIC_LOCATED_H
+#include "swift/Basic/Debug.h"
+#include "swift/Basic/LLVM.h"
+#include "swift/Basic/SourceLoc.h"
+
+namespace swift {
+
+/// A currency type for keeping track of items which were found in the source code.
+/// Several parts of the compiler need to keep track of a `SourceLoc` corresponding
+/// to an item, in case they need to report some diagnostics later. For example,
+/// the ClangImporter needs to keep track of where imports were originally written.
+/// Located makes it easy to do so while making the code more readable, compared to
+/// using `std::pair`.
+template<typename T>
+struct Located {
+
+  /// The main item whose source location is being tracked.
+  T Item;
+
+  /// The original source location from which the item was parsed.
+  SourceLoc Loc;
+
+  Located(): Item(), Loc() {}
```

Comment:
Hm, I wonder if it makes sense to have this constructor given that not every T can be default or value-initialized.  It might make more sense to provide an `explicit` constructor from a `T` or `const T&`, or default the `loc` in the other constructor and make it `explicit`.

Reply:
If this is the route you take, please also define an `explicit` move constructor.

Reply:
I think the reason for adding this was that `DenseMap` required this constructor for its values?

Reply:
We should give a DenseMapInfo specialization to avoid that...

Reply:
Hi, @CodaFi
As I understand correctly from code base we are making such specializations for Key:
```c++
namespace llvm {

template <typename T> struct DenseMapInfo;

template<typename T>
struct DenseMapInfo<swift::Located<T>> {

  static inline swift::Located<T> getEmptyKey() {
    return swift::Located<T>(T(), swift::SourceLoc());
  }

  static inline swift::Located<T> getTombstoneKey() {
    return swift::Located<T>(T(), swift::SourceLoc());
  }

  static unsigned getHashValue(const swift::Located<T> &LocatedVal) {
    return combineHashValue(DenseMapInfo<T>::getHashValue(LocatedVal.Item),
                            DenseMapInfo<swift::SourceLoc>::getHashValue(LocatedVal.Loc));
  }

  static bool isEqual(const swift::Located<T> &LHS, const swift::Located<T> &RHS) {
    return LHS == RHS;
  }
};
} // namespace llvm
```
But problem in DenseMap
```c++
value_type& FindAndConstruct(const KeyT &Key) {
    BucketT *TheBucket;
    if (LookupBucketFor(Key, TheBucket))
      return *TheBucket;

    return *InsertIntoBucket(TheBucket, Key);
  }

  ValueT &operator[](const KeyT &Key) {
    return FindAndConstruct(Key).second /**here**/;
  }
```
As I understand it tries to default initialize Value. And solution with DenseMapInfo specialization will not help in such case, where Located<T> is  Value. Is it right ?

Reply:
Tombstone and empty key requirements are not correctly implemented.  Take a look at how LLVM implements this for std::pair to get an idea of how this should look.  You need to delegate to the DenseMapInfo of T and the SourceLoc.